presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java

```diff
@@ -390,6 +360,51 @@ public String toString()
         return stateMachine.toString();
     }
 
+    private class StageTaskListener
+            implements StateChangeListener<TaskInfo>
+    {
+        private long previousMemory;
+
+        @Override
+        public void stateChanged(TaskInfo taskInfo)
+        {
+            long currentMemory = taskInfo.getStats().getMemoryReservation().toBytes();
+            long deltaMemoryInBytes = currentMemory - previousMemory;
```

Comment:
This needs to be made thread safe. `stateChanged` is invoked from different threads in `StateMachine.fireStateChanged`, possibly even concurrently